scaffolding-python/lib/scaffolding.sh

```diff
@@ -11,62 +11,215 @@
+    # Python with emit a non-zero return status if setup.py does not exist, or
+    # if the option passed is invalid.
+    # For valid but undefined values it will return "UNKNOWN", with the
+    # exception of `--version` for which it will return `0.0.0`.
+    if [[ $? -ne 0 ]] || [[ "$value" == "UNKNOWN" ]] || [[ "$value" == "0.0.0" ]]; then
+      return 5
+    else
+      printf "%s" "$value"
+    fi
+  fi
+}
+
+# Fetch and assign the pkg_upstream_url if it is not set in the plan, and is
+# defined in the setup.py.
+_scaffolding_python_setup_pkg_upstream_url() {
+  local url
+  url="$(_scaffolding_python_fetch_setup_info "url")"
+
+  if [[ $? -eq 0 ]]; then
+    build_line "Setting pkg_upstream_url=$url"
+    pkg_upstream_url=${pkg_upstream_url:-"$url"}
+  fi
+}
+
+# Fetch and assign the pkg_description if it is not set in the plan, and is
+# defined in the setup.py. Attempts to use `long_description`, and falls back
+# to description if `long_description` is undefined in the setup.py.
+_scaffolding_python_setup_pkg_description() {
+  local name
+  description="$(_scaffolding_python_fetch_setup_info "long-description" || _scaffolding_python_fetch_setup_info "description")"
+
+  if [[ $? -eq 0 ]]; then
+    build_line "Setting pkg_description=$description"
+    pkg_description=${pkg_description:-"$description"}
+  fi
 }
+scaffolding_python_venv() {
+  if [[ ! -f "$pkg_prefix/bin/activate" ]]; then
+    build_line "Creating package venv"
+    $pkg_scaffolding_python_venv_cmd \
+      --symlinks \
+      "$pkg_prefix"
+  fi
 
+  if source "$pkg_prefix/bin/activate"; then
+    build_line "Activated venv"
+  fi
+}
 
-# Recursively clean GOPATH and dependencies.
-do_default_clean() {
-  scaffolding_python_clean
+
+# Attempts to automatically set various pkg_* variables.
+scaffolding_python_prepare() {
+  _scaffolding_python_setup_pkg_description
+  _scaffolding_python_setup_pkg_upstream_url
 }
 
+# Removes bytecode and other stuff we do not need to package up.
 scaffolding_python_clean() {
-  pushd "$pkg_dirname"
-  if [[ -f "setup.py" ]]; then
+  if [[ $(_setup_py_exists) ]]; then
+    pushd "$CACHE_PATH" >/dev/null
     python setup.py clean --all
+    popd >/dev/null
   else
-    find . \( -type f -name "*.pyc" -or -name "*.pyo" \) -delete
+    find "$CACHE_PATH" \( -type f -name "*.pyc" -or -name "*.pyo" \) -delete
   fi
-  popd
 }
 
+# Run setup.py if it exists, otherwise return so pip can do the install.
 scaffolding_python_build() {
-  python setup.py build
+  scaffolding_python_venv
+  if [[ $(_setup_py_exists) ]]; then
+    build_line "Running setup.py"
+    "$pkg_prefix/bin/python" setup.py build --executable="$pkg_prefix/bin/python"
+  else
+    return 0
+  fi
 }
 
+# This can be called directly to install multiple pip packages.
+# Usage: `scaffolding_python_pip_install "mypackage "myversion"`
+# OR `scaffolding_python_pip_install "mypackage" "myversion" "--other-pip-arg"`
 scaffolding_python_pip_install() {
```

Comment:
We should respect the user's wish to do `pip install .` to allow downloading of py dependencies that have been defined.